server/botany/views.py

```diff
@@ -318,5 +319,35 @@ def api_submit(request):
     return JsonResponse({})
 
 
+def download_bots_code(request):
+    if datetime.now(timezone.utc) < settings.BOTANY_TOURNAMENT_CLOSE_AT:
+        return HttpResponseBadRequest(
+            "Unable to download bots while tournament is still in progress"
+        )
+
+    if not request.user.is_authenticated:
+        raise PermissionDenied
+
+    zip_buffer = get_active_bots()
+    response = HttpResponse(
+        zip_buffer.getvalue(), content_type="application/zip"
+    )
+    response["Content-Disposition"] = "attachment; filename=bots.zip"
+    return response
+
+
+@csrf_exempt
+def api_download_bots_code(request):
+    if datetime.now(timezone.utc) < settings.BOTANY_TOURNAMENT_CLOSE_AT:
+        return HttpResponseBadRequest(
+            "Unable to download bots while tournament is still in progress"
+        )
+
+    # check that user exists with given token
+    get_object_or_404(User, api_token=request.META["HTTP_AUTHORIZATION"])
+
+    return JsonResponse(get_active_bots_for_api(), safe=False)
```

Comment:
I did weigh up the two options and I'm not sure which is actually best, but my gut said converting some strings to json and then writing to files was somehow the less complex option. Please let me know if you have the opposite preference!

Reply:
There's not much in it, but to me your approach seems more complex because (a) there are two code paths that achieve much the same thing (make bot code available to an end user), and (b) the client code needs to manage writing to the file system, when it could just call `.extract()` on a zipfile.

Reply:
> I've no idea if that would ever be of use anyway
Sounds like a good idea not to have it then :)

> Would you recommend removing the /api/download-bots/ URL & view entirely and just fetching the /download-bots/ URL from the client?
I'd have two separate URLs and two separate views, where these views just handle authentication (since this will differ between requests from the browser and requests from the client), and then call a function called `_download_bots()` that will actually return the `HttpResponse`.

Reply:
Good point re auth. Okay, will sort this :)

Reply:
I've done this over a few commits.

The most recent commit, 7b77fd7, does introduce a patch, but hopefully it doesn't cause any concern regarding maintainability in this case. Since all we want to do is check that `download_bots_code` and `api_download_bots_code` return the result of a call to `_download_bots_code`, without modifying it, I'm hoping this is an ideal time to use a patch?

Reply:
I wouldn't personally use a patch here, because if we ever make a change so that the behaviour of `views.download_bots_code` is unchanged (as in, it still returns a request with a zip file) but the implementation does change so as to not use `views._download_bots_code`, the test will fail, but this failure won't tell us anything useful.

(Not that I plan to change the implementation... it's just that I don't think patching is useful here.)

I appreciate your desire to reduce duplication in the tests, but a different approach would be to write a function or method called something like `assertZipContains`, and then to call that from `test_download_bots_code_view` and `test_api_download_bots_code_view`.  That way you could also remove `test_download_bots_code_helper_function` -- I'd personally not test that directly, as it's an implementation detail and not part of the external interface.

But I'm happy to keep it in for the time being -- there's been a lot of to-ing and fro-ing.

Apart from this, do you think we're done with this PR?

Reply:
I was hoping using "Return value of _download_bots_code" would make any test failure on this a little clearer on what has gone wrong.

I committed this separately on purpose - I assume that's relatively painless to remove should you want to.

I do like the idea of a assertZipContains, but I'll leave that for later.

Yes, I'm happy enough with this. Sorry if it's been more hassle than it needed to be :\